src/components/backend-ai-signup.ts

```diff
@@ -264,31 +308,145 @@ export default class BackendAiSignup extends BackendAIPage {
+            patternMismatch: !nativeValidity.valid
+          }
+        } else {
+          passwordInput.validationMessage = _text('signup.PasswordInputRequired');
+          return {
+            valid: nativeValidity.valid,
+            customError: !nativeValidity.valid
+          }
+        }
+      }
+    }
+  }
+
+  _validatePassword2() {
+    const password2Input = this.shadowRoot.querySelector('#id_password2');
+    password2Input.validityTransform = (newValue, nativeValidity) => {
+      if (!nativeValidity) {
+        let isValid: boolean = nativeValidity;
+        return {
+          valid: !isValid,
+          customError: isValid
+        }
+      }
+      if (!nativeValidity.valid) {
+        if (nativeValidity.tooShort) {
+          password2Input.validationMessage = _text('signup.PasswordInvalid');
+          return {
+            valid: nativeValidity.valid,
+            tooShort: !nativeValidity.valid
+          } 
+        } else if (nativeValidity.patternMismatch) {
+          password2Input.validationMessage = _text('signup.PasswordInvalid');
+          return {
+            valid: nativeValidity.valid,
+            patternMismatch: !nativeValidity.valid
+          }
+        }
+      } else {
+        // custom validation for password input match
+        const passwordInput = this.shadowRoot.querySelector('#id_password1');
+        let isMatched = (passwordInput.value === password2Input.value);
+        if (!isMatched) {
+          password2Input.validationMessage = _text('signup.PasswordNotMatched');         
+        }
+        return {
+          valid: isMatched,
+          customError: !isMatched
+        }
+      }
+    }
+  }
+
+  _validatePassword() {
+    this._validatePassword1();
+    this._validatePassword2();
+  }
+
   render() {
     // language=HTML
     return html`
       <backend-ai-dialog id="signup-panel" fixed blockscrolling persistent disablefocustrap>
         <span slot="title">${_t("signup.SignupBETA")}</span>
         <div slot="content">
           <mwc-textfield type="text" name="user_email" id="id_user_email" maxlength="50" autofocus
-                       label="${_t("signup.E-mail")}" value="${this.user_email}"
-                       @change="${() => this._clear_info()}"></mwc-textfield>
+                       label="${_t("signup.E-mail")}" validateOnInitialRender
+                       pattern="^[^\s@]+@[^\s@]+\.[^\s@]+$" @change="${this._validateEmail}" 
+                       validationMessage="${_t("signup.EmailInputRequired")}"
+                       value="${this.user_email}" required></mwc-textfield>
           <mwc-textfield type="text" name="user_name" id="id_user_name" maxlength="30"
-                       label="${_t("signup.UserName")}" value="${this.user_name}"></mwc-textfield>
+                       label="${_t("signup.UserName")}" value="${this.user_name}"
+                       validationMessage="${_t("signup.UserNameInputRequired")}" required></mwc-textfield>
```

Comment:
In the server-side, `username` is actually not a required field. If `username` is missing, manager will use `email` as a `username`.  But, if you think username should be required from the client-side, I have no further objection.

Reply:
I'm all for that plan. I'll change `username` field to option from required.